semantic-scope-graph/src/Stack/Path.hs

```diff
@@ -0,0 +1,130 @@
+{-# LANGUAGE NamedFieldPuns #-}
+{-# LANGUAGE OverloadedLists #-}
+module Stack.Path
+  ( Path (..)
+  , Edge (..)
+  , StartingSize (..)
+  , PathInvariantError (..)
+  , checkEdgeInvariants
+  , checkNodeInvariants
+  , Validity (..)
+  , validity
+  , Completion (..)
+  , completion
+  , isIncremental
+  ) where
+
+
+import Data.Functor.Tagged
+import Data.Monoid
+import Data.Text (Text)
+import Stack.Graph (Node (..), Symbol)
+import Data.Semigroup (sconcat)
+import Data.Sequence (Seq (..))
+
+data Path = Path
+  { startingNode           :: Tagged Node
+  , endingNode             :: Tagged Node
+  , edges                  :: Seq Edge
+  , startingSymbolStack    :: [Symbol]
+  , endingSymbolStack      :: [Symbol]
+  , startingScopeStackSize :: StartingSize
+  , endingScopeStack       :: [Tag]
+  } deriving (Eq, Show)
+
+data Edge = Edge
+  { sourceNode :: Tagged Node
+  , sinkNode   :: Tagged Node
+  , label      :: Text
+  } deriving (Eq, Show)
+
+data StartingSize
+  = Zero
+  | One
+  deriving (Eq, Show)
+
+data PathInvariantError
+  = ExpectedEqual (Tagged Node) (Tagged Node)
+  | BadStartingNode (Tagged Node)
+  | BadEndingNode (Tagged Node)
+    deriving (Eq, Show)
+
+-- | If a path's edges list is empty, then its starting node must be
+-- the same as its ending node. If a path's edges list is nonempty,
+-- then the starting node of the path must be the same as the source
+-- node of the first edge in the path, and the ending node of the path
+-- must be the same as the sink node of the last edge in the path.
+checkEdgeInvariants :: Path -> Maybe PathInvariantError
+checkEdgeInvariants Path{ startingNode, endingNode, edges }
+  = let
+      check :: Tagged Node -> Tagged Node -> First PathInvariantError
+      check a b = if a /= b then pure (ExpectedEqual a b) else mempty
+    in getFirst $ case edges of
+         Empty
+           -> check startingNode endingNode
+         Edge { sourceNode, sinkNode } :<| Empty
+           -> check startingNode sourceNode <> check endingNode sinkNode
+         Edge { sourceNode  } :<| (_ :|> Edge { sinkNode })
+           -> check startingNode sourceNode <> check endingNode sinkNode
+
+-- | The starting node of a path must be the root node, an exported
+-- scope node, or a reference node. The ending node of a path must be
+-- the root node, a jump to scope node, a resolved jump to scope node,
```

Comment:
> resolved jump to scope node

Gah!  This is a typo in the design doc, "resolved jump to scope node" isn't a type of node anymore.  That clause should be removed.